src/enclave/enclave.h

```diff
@@ -146,29 +153,33 @@ namespace enclave
         *writer_factory, network, rpcsessions, share_manager, curve_id);
 
       LOG_TRACE_FMT("Creating context");
-      context = std::make_unique<NodeContext>();
+      context = std::make_unique<NodeContext>(node->get_node_id());
       context->historical_state_cache =
         std::make_shared<ccf::historical::StateCache>(
           *network.tables,
           network.ledger_secrets,
           writer_factory->create_writer_to_outside());
-      context->node_state = node.get();
       context->indexer = std::make_shared<ccf::indexing::Indexer>(
         std::make_shared<ccf::indexing::HistoricalTransactionFetcher>(
           context->historical_state_cache));
       context->lfs_access = std::make_unique<ccf::indexing::EnclaveLFSAccess>(
         writer_factory->create_writer_to_outside());
+      context->host_processes = std::make_unique<ccf::HostProcesses>(*node);
 
       LOG_TRACE_FMT("Creating RPC actors / ffi");
+      node_operation = std::make_unique<ccf::NodeOperation>(*node);
+      gov_effects = std::make_unique<ccf::GovernanceEffects>(*node);
+
       rpc_map->register_frontend<ccf::ActorsType::members>(
         std::make_unique<ccf::MemberRpcFrontend>(
-          network, *context, share_manager));
+          network, *context, share_manager, *gov_effects, *node_operation));
```

Comment:
This section highlights the core functional change here. `MemberRpcFrontend` used to take an `AbstractNodeContext`, on which it could call things like `context.get_node_state().is_reading_private_ledger()` and `context.get_node_state().rekey_ledger()`. But that same `AbstractNodeContext` was also exposed to app frontends, for the few bits of functionality they also needed (in practice I think that's just "getting the current node's ID"). This means apps could call those functions too, where the former is _maybe_ safe and the latter is _definitely not!_ So we remove that functionality from the shared `AbstractNodeContext`, which no longer has a `get_node_state()`, and instead move those methods to subsystem-grouped interfaces like `GovernanceEffects` or `NodeOperation`, and pass those only to the frontends under our control.

Reply:
And an addendum to this with recent changes: In the initial version of this PR we had to add something like `*gov_effects, *node_operation` to every constructor of a `MemberRpcFrontend` (including tests that probably don't use them at all!). Now those are _maybe_ available from the existing `context` argument - the constructor doesn't bloat, the dependency on this subsystem can be moved closer to where it's actually needed rather than being widely intrusive.